### ISSUE TITLE
cannot load shared library dllbigstringaf_stubs

### ISSUE THREAD

**Issue by eeelllccc** [tweag/opam-nix#93]
**Describe the bug**
```
Fatal error: cannot load shared library dllbigstringaf_stubs
Reason: dllbigstringaf_stubs.so: cannot open shared object file: No such file or directory
``` 
As a runtime error only when the binary has been built from the flake using opam-nix. When I build on my mac (m1 pro) with dune there are no runtime errors.

**To Reproduce**
`nix build https://github.com/edchapman88/getting/archive/refs/tags/v0.2.7.tar.gz`
and then run the binary with the command: `getting`

**Expected behavior**
No runtime error.

**Environment**
 - OS name + version: NixOS 24.05.20240916.086b448 (Uakari) with `system.stateVersion = 23.11` in configuration.nix

**Additional context**
I noticed the other [issue](https://github.com/tweag/opam-nix/issues/9) about dll linking, and so thought it might be opam-nix related.

**Comment by eeelllccc** [comment]
I'm still trying to get to the bottom of this (and also figure out whether it is problem that can be address by opam-nix, or perhaps somewhere else in the stack). 

The problem seems to be an incompatibility between building a nix package with opam-nix, and having dune project dependencies that use [foreign build sandboxing](https://dune.readthedocs.io/en/latest/foreign-code.html#foreign-sandboxing) to build foreign libraries (e.g. C libs).

I've tried using static and dynamic build with opam-nix (e.g. setting
```        scope =
          on.buildDuneProject { pkgs = pkgs.pkgsStatic; } package ./. { ocaml-base-compiler = "*"; };
```
vs
```
        scope =
          on.buildDuneProject { } package ./. { ocaml-base-compiler = "*"; };
```

When using static builds, the build fails with:
```
> File "vendor/dune", line 9, characters 0-251:
       >  9 | (rule
       > 10 |  (targets libyaml_c_stubs%{ext_lib} dllyaml_c_stubs%{ext_dll})
       > 11 |  (deps api%{ext_obj} emitter%{ext_obj} loader%{ext_obj} parser%{ext_obj}
       > 12 |    reader%{ext_obj} scanner%{ext_obj} writer%{ext_obj})
       > 13 |  (action
       > 14 |   (run ocamlmklib -o yaml_c_stubs %{deps})))
       > Error: Rule failed to generate the following targets:
       > - vendor/dllyaml_c_stubs.so
```
In other words, the `ocamlmklib` command fails to build the `dllyaml_c_stubs` target.

And when using dynamically linked builds, the build succeeds but the executable fails at runtime with:
```
Fatal error: cannot load shared library dllbase_internalhash_types_stubs
Reason: dllbase_internalhash_types_stubs.so: cannot open shared object file: No such file or directory
```
E.g. in this case the first dll to be called was from a different opam dependency that probably also uses foreign build sandboxing

Not sure how to proceed.

**Comment by balsoft** [comment]
Hey, sorry for the delay with the reply.

I've never seen issues like this before with `opam-nix`, but my first thought is that it just fails to find the library in its RPATH? Perhaps some combination of `LD_LIBRARY_PATH` will help it along.

**Comment by balsoft** [comment]
I don't have the time to figure out what this project is supposed to do, but I could get it to run by modifying the `LD_LIBRARY_PATH`:

```
❯ LD_LIBRARY_PATH=$(for reference in $(nix-store -q --references ./result); do find $reference -name 'stublibs'; done | tr '\n' ':' | sed 's/:$//') ./result/bin/getting -ignore-fd-limit balsoft.ru
https://balsoft.ru
1
```

**Comment by balsoft** [comment]
Perhaps you could use `patchelf --add-rpath` to add all those `stublibs` directories the the RPATH of your binary, e.g. in `postFixupPhase`. I find it weird that it doesn't do this automagically, I'll take a look later.

**Comment by eeelllccc** [comment]
Thanks for helping find the `LD_LIBRARY_PATH` workaround, that works for me too. 

I'm struggling to get patchelf to solve the problem in the flake instead. Tried:
```
postFixupPhase = ''
${pkgs.patchelf}/bin/patchelf --add-rpath $out/lib $out/bin/getting
'';
```
which doesn't help. I'm very new to Nix, not really sure how to get a handle on the 'will-be' nix-store paths in the postFixupPhase of the derivation itself?

**Comment by balsoft** [comment]
The reason it doesn't work automagically (I think) is because of this:

```
❯ file result/bin/getting
result/bin/getting: a /nix/store/bnavs25x0zqk0k2sgm1nvb65nj0asgr9-ocaml-base-compiler-5.1.1_rc1/bin/ocamlrun script executable (binary data)
```

I think this is due to `(modes byte)` in `bin/dune`, which forces it to link it as bytecode instead of ELF. Thus, the standard linking flags do not apply and there's no RPATH in the resulting binary.

This `flake.nix` on the main branch of the project makes it succeed and run:

```nix
{
  inputs = {
    opam-nix.url = "github:tweag/opam-nix";
    flake-utils.url = "github:numtide/flake-utils";
    nixpkgs.follows = "opam-nix/nixpkgs";
  };
  outputs = { self, flake-utils, opam-nix, nixpkgs }@inputs:
    # Don't forget to put the package name instead of `throw':
    let package = "getting";
    in flake-utils.lib.eachDefaultSystem (system:
      let
        pkgs = nixpkgs.legacyPackages.${system};
        on = opam-nix.lib.${system};
        scope =
          on.buildDuneProject { } package ./. { ocaml-base-compiler = "*"; };
        overlay = final: prev:
          {
            getting = prev.getting.overrideAttrs (oa: {
              nativeBuildInputs = oa.nativeBuildInputs ++ [ pkgs.makeWrapper ];
              postInstall = ''
                wrapProgram "$out/bin/getting" --prefix "LD_LIBRARY_PATH" ":" "$CAML_LD_LIBRARY_PATH"
              '';
            });

          };
      in {
        legacyPackages = scope.overrideScope' overlay;

        packages.default = self.legacyPackages.${system}.${package};
      });
}
```

But, if possible, I would recommend to just remove `(modes byte)` and let it link natively, which also just fixes the issue for me.

**Comment by balsoft** [comment]
I'll think if it makes sense to explicitly support bytecode executables in `opam-nix`. If it does, I'll add a wrapper that does the `wrapProgram` automatically. Keeping this issue open for now.

**Comment by balsoft** [comment]
Oh, and BTW, the reason your overlay didn't do anything is because you tried to set package attributes, but instead (accidentally) added new packages with those names to scope. Compare correct (even if non-working)

```
        overlay = final: prev:
          {
            getting = prev.getting.overrideAttrs (oa: {
               postFixupPhase = ''
                 ${pkgs.patchelf}/bin/patchelf --add-rpath $out/lib $out/bin/getting
               '';
            });
          };
```

With incorrect

```
        overlay = final: prev:
          {
               postFixupPhase = ''
                 ${pkgs.patchelf}/bin/patchelf --add-rpath $out/lib $out/bin/getting
               '';
          };
```

Which just adds a very weird "package" called `postFixupPhase` to your scope, and doesn't change the `getting` package at all.

You can read more about overlays here: https://nixos.org/manual/nixpkgs/stable/#sec-overlays-definition and https://nixos.wiki/wiki/Overlays

**Comment by eeelllccc** [comment]
Thank you so much for the help! 

**For the benefit of anyone else reading this thread - I introduced `(modes byte)` in order to use the ocaml debugger, so watch out for this unintended consequence. `(modes byte exe)` builds both an elf and a byte-code program, which won't break the nix native linking.**

Re: supporting bytecode executables in `opam-nix`, it's certainly not necessary in my use case, although perhaps it might be worth added a sentence to the `opam-nix` docs mentioning this problem incase there are any other beginners that run into this too.

Also explains the output I was getting from patchelf (when i initially had the overlay correctly implemented!) saying `not an elf file`!

Really grateful for the package btw, wouldn't be able to use ocaml on this project without it, which would have been a shame.

**Comment by balsoft** [comment]
After some consideration, I don't think adding bytecode support is worth the effort. I can't find many use-cases for it that would overlap with use-cases for opam-nix, and it would require quite a lot of new code.
